node/core/proposer/src/lib.rs

```diff
@@ -0,0 +1,323 @@
+	transaction_pool: Arc<TxPool>,
+	/// phantom member to pin the `Backend` type.
+	_phantom: PhantomData<Backend>,
+}
+
+impl<TxPool, Backend, Client> ProposerFactory<TxPool, Backend, Client> {
+	pub fn new(client: Arc<Client>, transaction_pool: Arc<TxPool>) -> Self {
+		ProposerFactory {
+			client,
+			transaction_pool,
+			_phantom: PhantomData,
+		}
+	}
+}
+
+impl<Backend, Block, Client, TxPool> ProposerFactory<TxPool, Backend, Client>
+where
+	TxPool: TransactionPool<Block = Block> + 'static,
+	Backend: backend::Backend<Block> + Send + Sync + 'static,
+	Block: BlockT,
+	Client: BlockBuilderProvider<Backend, Block, Client>
+		+ HeaderBackend<Block>
+		+ ProvideRuntimeApi<Block>
+		+ Send
+		+ Sync
+		+ 'static,
+	Client::Api: ApiExt<Block, StateBackend = backend::StateBackendFor<Backend, Block>>
+		+ BlockBuilderApi<Block, Error = sp_blockchain::Error>,
+{
+	pub fn init_with_now(
+		&mut self,
+		parent_header: &<Block as BlockT>::Header,
+		now: Box<dyn Fn() -> time::Instant + Send + Sync>,
+	) -> Proposer<TxPool, Backend, Client, Block> {
+		let parent_hash = parent_header.hash();
+
+		let id = BlockId::hash(parent_hash);
+
+		info!(
+			"🙌 Starting consensus session on top of parent {:?}",
+			parent_hash
+		);
+
+		let proposer = Proposer(Arc::new(ProposerInner {
+			client: self.client.clone(),
+			parent_hash,
+			parent_id: id,
+			parent_number: *parent_header.number(),
+			transaction_pool: self.transaction_pool.clone(),
+			now,
+			_phantom: PhantomData,
+		}));
+
+		proposer
+	}
+}
+
+impl<TxPool, Backend, Client, Block> sp_consensus::Environment<Block>
+	for ProposerFactory<TxPool, Backend, Client>
+where
+	TxPool: TransactionPool<Block = Block> + 'static,
+	Backend: backend::Backend<Block> + Send + Sync + 'static,
+	Block: BlockT,
+	Client: BlockBuilderProvider<Backend, Block, Client>
+		+ HeaderBackend<Block>
+		+ ProvideRuntimeApi<Block>
+		+ Send
+		+ Sync
+		+ 'static,
+	Client::Api: ApiExt<Block, StateBackend = backend::StateBackendFor<Backend, Block>>
+		+ BlockBuilderApi<Block, Error = sp_blockchain::Error>,
+{
+	type CreateProposer = future::Ready<Result<Self::Proposer, Self::Error>>;
+	type Proposer = Proposer<TxPool, Backend, Client, Block>;
+	type Error = sp_blockchain::Error;
+
+	fn init(&mut self, parent_header: &<Block as BlockT>::Header) -> Self::CreateProposer {
+		future::ready(Ok(
+			self.init_with_now(parent_header, Box::new(time::Instant::now))
+		))
+	}
+}
+
+pub struct Proposer<TxPool, Backend, Client, Block: BlockT>(
+	Arc<ProposerInner<TxPool, Backend, Client, Block>>,
+);
+
+struct ProposerInner<TxPool, Backend, Client, Block: BlockT> {
+	client: Arc<Client>,
+	parent_hash: <Block as BlockT>::Hash,
+	parent_id: BlockId<Block>,
+	parent_number: <<Block as BlockT>::Header as HeaderT>::Number,
+	transaction_pool: Arc<TxPool>,
+	now: Box<dyn Fn() -> time::Instant + Send + Sync>,
+	_phantom: PhantomData<Backend>,
+}
+
+impl<TxPool, Backend, Client, Block> sp_consensus::Proposer<Block>
```

Comment:
@rphmeier What is the reason to duplicate the logic again here in Polkadot?

Why not just wrap the Substrate `Proposer`?

Reply:
@bkchr Yeah, @coriolinus and I went over that in DM before. My guess is that the next push will contain a wrapper similar to what is already in `polkadot-validation` as opposed to duplication of logic.